jax/flatten_util.py

```diff
@@ -26,18 +30,36 @@ def ravel_pytree(pytree):
   """Ravel (i.e. flatten) a pytree of arrays down to a 1D array.
 
   Args:
-    pytree: a pytree to ravel.
+    pytree: a pytree of arrays and scalars to ravel.
 
   Returns:
     A pair where the first element is a 1D array representing the flattened and
-    concatenated leaf values, and the second element is a callable for
-    unflattening a 1D vector of the same length back to a pytree of of the same
-    structure as the input ``pytree``.
+    concatenated leaf values, with dtype determined by promoting the dtypes of
+    leaf values, and the second element is a callable for unflattening a 1D
+    vector of the same length back to a pytree of of the same structure as the
+    input ``pytree``.
+
+  For details on dtype promotion, see
+  https://jax.readthedocs.io/en/latest/type_promotion.html.
   """
   leaves, treedef = tree_flatten(pytree)
-  flat, unravel_list = vjp(_ravel_list, *leaves)
+  flat, unravel_list = _ravel_list(leaves)
   unravel_pytree = lambda flat: tree_unflatten(treedef, unravel_list(flat))
   return flat, unravel_pytree
 
-def _ravel_list(*lst):
-  return jnp.concatenate([jnp.ravel(elt) for elt in lst]) if lst else jnp.array([])
+def _ravel_list(lst):
+  from_dtypes = [dtypes.dtype(l) for l in lst]
```

Comment:
Is it important to preserve weak type information in the outputs? If so, you could use `dtype._dtype_and_weaktype` for each entry, and use `to_dtype = jnp.result_type(*leaves)`

Reply:
I'm not sure. Actually you should tell me, since you are the weak type expert :)

My guess is it is not important.

Reply:
My guess is it's not important because AIUI weak types are primarily useful for handling Python scalar literals in user code in a reasonable way. But we don't have to think about the round-tripped output as a Python scalar literal.

Reply:
It sounds reasonable to ignore weak types in unravelling. But for the ravelled dtype, it may be important. For example, should we expect `ravel_pytree(jnp.bfloat16(0.0), 1.0))` ravel to `bfloat16` or to `float32`?

Reply:
Good thinking! What do you recommend as the answer to that question?

Reply:
Actually, I'd really appreciate if you could recommend a specific change. Otherwise I'm happy with landing it as-is and not worrying about it too much :)

Reply:
I'm not entirely familiar with where this is intended to be used, but it looks like weak types are ignored in the current implementation, so it seems reasonable to keep doing that.

Reply:
SG!

Thanks for the advice. Anything else blocking this, or LGTY?